File: src/main/java/com/epri/dss/parser/impl/RPNCalcImpl.java
package com.epri.dss.parser.impl;

import com.epri.dss.parser.RPNCalc;

public class RPNCalcImpl implements RPNCalc {

	private static final int MaxStackSize = 10;
	
	private static final double DegToRad = 3.14159265359 / 180.0;
	private static final double RadToDeg = 1.0 / DegToRad;

	private double[] Stack = new double[MaxStackSize];

	public RPNCalcImpl() {
		for (int i = 0; i < MaxStackSize; i++) 
			Stack[i] = 0.0;
	}

	public void aCosDeg() {
		Stack[0] = RadToDeg * Math.acos(Stack[0]);
	}

	public void add() {
		Stack[1] = Stack[0] + Stack[1];
		rollDn();
	}

	public void aSinDeg() {
		Stack[0] = RadToDeg * Math.asin(Stack[0]);
	}

	public void aTanDeg() {
		Stack[0] = RadToDeg * Math.atan(Stack[0]);
	}

	public void aTan2Deg() {
		Stack[1] = RadToDeg * Math.atan2(Stack[1], Stack[0]);
		rollDn();
	}

	public void cosDeg() {
		Stack[0] = Math.cos(DegToRad * Stack[0]);
	}

	public void divide() {
		Stack[1] = Stack[1] / Stack[0];
		rollDn();
	}

	public double getX() {
		return Stack[0];
	}

	public double getY() {
		return Stack[1];
	}

	public double getZ() {
		return Stack[2];
	}

	public void multiply() {
		Stack[1] = Stack[1] * Stack[0];
		rollDn();
	}

	public void rollDn() {
		for (int i = 1; i < MaxStackSize; i++) 
			Stack[i - 1] = Stack[i];
	}

	public void rollUp() {
		for (int i = 1; i < MaxStackSize; i++) 
			Stack[i] = Stack[i - 1];
		
	}

	public void setX(double x) {
		rollUp();
		Stack[0] = x;
	}

	public void setY(double y) {
		Stack[1] = y;
	}

	public void setZ(double z) {
		Stack[2] = z;
	}

	public void sinDeg() {
		Stack[0] = Math.sin(DegToRad * Stack[0]);
	}

	public void sqrt() {
		Stack[0] = Math.sqrt(Stack[0]);
	}

	public void square() {
		Stack[0] = Math.pow(Stack[0], 2);
	}

	public void subtract() {
		Stack[1] = Stack[1] - Stack[0];
		rollDn();
	}
	
	public void swapXY() {
		double Temp = Stack[0];
		Stack[0] = Stack[1];
		Stack[1] = Temp;
	}

	public void tanDeg() {
		Stack[0] = Math.tan(DegToRad * Stack[0]);
	}

	public void yToTheXPower() {
		Stack[1] = Math.pow(Stack[1], Stack[0]);
		rollDn();
	}

	public void enterPi() {
		rollUp();
		Stack[0] = Math.PI;
	}

	public void eToTheX() {
		Stack[0] = Math.exp(Stack[0]);
	}

	public void natLog() {
		Stack[0] = Math.log(Stack[0]);
	}

	public void tenLog() {
		Stack[0] = Math.log10(Stack[0]);
	}

	public void inv() {
		Stack[0] = 1.0 / Stack[0];
	}

}

File: pydss/parser/RPNCalcImpl.py
from dss.parser.RPNCalc import (RPNCalc,)


class RPNCalcImpl(object, RPNCalc):
    MaxStackSize = 10
    DegToRad = 3.14159265359 / 180.0
    RadToDeg = 1.0 / DegToRad
    stack = [None] * MaxStackSize

    def __init__(self):
        _0 = True
        i = 0
        while True:
            if _0 is True:
                _0 = False
            else:
                i += 1
            if not (i < self.MaxStackSize):
                break
            self.stack[i] = 0.0

    def aCosDeg(self):
        self.stack[0] = self.RadToDeg * self.Math.acos(self.stack[0])

    def add(self):
        self.stack[1] = self.stack[0] + self.stack[1]
        self.rollDn()

    def aSinDeg(self):
        self.stack[0] = self.RadToDeg * self.Math.asin(self.stack[0])

    def aTanDeg(self):
        self.stack[0] = self.RadToDeg * self.Math.atan(self.stack[0])

    def aTan2Deg(self):
        self.stack[1] = self.RadToDeg * self.Math.atan2(self.stack[1], self.stack[0])
        self.rollDn()

    def cosDeg(self):
        self.stack[0] = self.Math.cos(self.DegToRad * self.stack[0])

    def divide(self):
        self.stack[1] = self.stack[1] / self.stack[0]
        self.rollDn()

    def getX(self):
        return self.stack[0]

    def getY(self):
        return self.stack[1]

    def getZ(self):
        return self.stack[2]

    def multiply(self):
        self.stack[1] = self.stack[1] * self.stack[0]
        self.rollDn()

    def rollDn(self):
        _0 = True
        i = 1
        while True:
            if _0 is True:
                _0 = False
            else:
                i += 1
            if not (i < self.MaxStackSize):
                break
            self.stack[i - 1] = self.stack[i]

    def rollUp(self):
        _0 = True
        i = 1
        while True:
            if _0 is True:
                _0 = False
            else:
                i += 1
            if not (i < self.MaxStackSize):
                break
            self.stack[i] = self.stack[i - 1]

    def setX(self, x):
        self.rollUp()
        self.stack[0] = x

    def setY(self, y):
        self.stack[1] = y

    def setZ(self, z):
        self.stack[2] = z

    def sinDeg(self):
        self.stack[0] = self.Math.sin(self.DegToRad * self.stack[0])

    def sqrt(self):
        self.stack[0] = self.Math.sqrt(self.stack[0])

    def square(self):
        self.stack[0] = self.Math.pow(self.stack[0], 2)

    def subtract(self):
        self.stack[1] = self.stack[1] - self.stack[0]
        self.rollDn()

    def swapXY(self):
        Temp = self.stack[0]
        self.stack[0] = self.stack[1]
        self.stack[1] = Temp

    def tanDeg(self):
        self.stack[0] = self.Math.tan(self.DegToRad * self.stack[0])

    def yToTheXPower(self):
        self.stack[1] = self.Math.pow(self.stack[1], self.stack[0])
        self.rollDn()

    def enterPi(self):
        self.rollUp()
        self.stack[0] = self.Math.PI

    def eToTheX(self):
        self.stack[0] = self.Math.exp(self.stack[0])

    def natLog(self):
        self.stack[0] = self.Math.log(self.stack[0])

    def tenLog(self):
        self.stack[0] = self.Math.log10(self.stack[0])

    def inv(self):
        self.stack[0] = 1.0 / self.stack[0]
